_posts/2024-08-28-Data-Prepper-2.9.0-is-ready-for-download.md

```diff
@@ -0,0 +1,120 @@
+You can define two sinks for these and define your routes.
+But, what if you receive events that do not match?
+The sample pipeline below shows you how you could handle events that do not match either the front-end or back-end routes.
+
+```
+routes:
+  - frontend: '/service == "front-end"`
+  - backend: '/service == "back-end"`
+sink:
+ - opensearch:
+      routes:
+         - front-end
+ - opensearch:
+      routes:
+         - back-end
+  - opensearch:
+       routes:
+          - _default
+```
+
+## Performance
+
+The Data Prepper maintainers have been working toward improving the performance of Data Prepper.
+This release includes a number of internal improvements that speed up processing for many processors.
+You don't have to do anything other than update your version to get these improvements.
+
+Data Prepper 2.9 also offers some new features that you can use to help reduce out-of-memory errors or circuit breaker trips.
+Many pipelines involve taking some source data in the form of a string and then extracting that data into a structure.
+Some examples of these are `grok` and `parse_json`.
+When you use these processors, you more than double the size of each event you are processing.
+As the events flowing through the system are the largest part of memory usage, this will greatly increase your memory requirements.
+
+Many pipeline authors will use these processors and then in a second processor remove the source data.
+This is good, but it doesn't always make the available available for garbage collection when you need it.
+The reason is that Data Prepper pipelines operate on batches of data.
+As these batches of data move through the pipeline, the pipeline will expand the memory usage in one processor and then attempt to reduce it in the next.
+Because these happen in batches, Data Prepper may expand many thousands of events, before starting to remove the source data.
+
+See the following example pipeline which runs `grok` and then `delete_entries`.
+With the configured `batch_size` of 100,000, Data Prepper will expand 100,000 events before deleting the messages.
+
+```
+my-pipeline:
+  buffer:
+    bounded_blocking:
+      batch_size: 100000
+  processor:
+    - grok:
+        match:
+          message: ["..."]
+    - delete_entries:
+        with_keys: ["message"]
+```
+
+To help with this memory usage issue, Data Prepper now provides a `delete_source` flag on some of these processors, including `grok` and `parse_json`.
+
+Going back to the example above, you could both simplify the pipeline and reduce the amount of memory used in between processors.
+
+```
+my-pipeline:
+  buffer:
+    bounded_blocking:
+      batch_size: 100000
+  processor:
+    - grok:
+        match:
+          message: ["..."]
+        delete_source: true
+```
+
+When you see this pattern of deleting the source in a separate processor, migrate your pipeline to start using `delete_source` instead to improve your overall memory usage.
+
```

Comment:
Line above: "migrate your pipeline to start using" => "configure your pipeline to use"?

Reply:
This looks good. I took it.